ArduPlane: add AC_Fence  for QGC4.0 Geofence Support

joshanne opened the pull request:
PR based on #13700 - see further discussion there

- Moves Plane from `Geofence` to `AC_Fence` for Fence support in QGC.
- `geofence` capability has been dropped in favour of `AC_Fence`. Some assumptions have been made.
- Adds floor breaching to `AC_Fence`
- Adds an autotest for testing Ceiling and Floor breaches. Issue in existing geofence implementation would be carried across (#5544)
- Existing autotests for `geofence` work, mostly untouched.
- Discussion to be had around mav status present/enabled bits.

Requires:
- Merge https://github.com/ArduPilot/ardupilot/pull/15568
- Documentation: https://github.com/ArduPilot/ardupilot_wiki/pull/3349

Todo:
- [x] minimum altitude fence breach - _autotests have been added_
- [x] ensure we can auto-enable the fence when vehicle armed - _autotests have been added_
- [x] should disallow arming if we are outside the fence when we auto-enable - _autotests have been added_
- [x] need tests around MAV_CMD_DO_FENCE_ENABLE
- [x] fence_ret_rally works - _autotests have been added._
- [x] need tests for what happens when fences are disabled while we're breached and undertaking an action - _autotests have been added_

joshanne commented:
Some tasks that remain outstanding:

- [ ] How should we handle AC_Fence or Geofence enabling? Currently set up with include guards.

peterbarker commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -245,7 +249,12 @@ void Plane::update_logging2(void)
 {
     // perform AFS failsafe checks
-    afs.check(failsafe.last_heartbeat_ms, geofence_breached(), failsafe.AFS_last_valid_rc_ms);
+#if AC_FENCE
```
By convention we always compare this against `== ENABLED`.  It was like that when I got here.

joshanne replied:
Ack - of all the places I added that check, I was bound to miss one.

peterbarker commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -414,12 +414,12 @@ bool GCS_MAVLINK_Plane::try_send_message(enum ap_message id)
         break;
 
-    case MSG_FENCE_STATUS:
 #if GEOFENCE_ENABLED == ENABLED
```
Not sending this under the new code?

joshanne replied:
This is handled by `GCS_MAVLink/GCS_Common.cpp:4549`

peterbarker reviewed (changes requested):
Started to review here, but retaining the existing geofence code is confusing the issue.

Why retain the existing geofence implementation?  We don't carry dead code, so....  Also, it seems to be half-and-half with retaining geofence functionality.

peterbarker commented:
Oh, also, alt_min may be <-100 for Subs :-)

peterbarker commented:
.... also, please squash your commits up into logical changes.  There are patches in here which are definitely "fixups" of previous commits.  I do realise this is WIP :-)

joshanne commented:
> Why retain the existing geofence implementation? We don't carry dead code, so.... Also, it seems to be half-and-half with retaining geofence functionality.

Fair call - the intention was more for the review process and an ability to swap between the two versions, or for those that want to maintain their existing `geofence` version (hence the separate ENABLE flags). Obviously, they wouldn't be compatible together. Happy to rip out the dead code, and move to an entirely `AC_Fence` model. 

> Oh, also, alt_min may be <-100 for Subs :-)

Good call - still trying to piece together how the documentation/param parsing works.
Note: There's also argument for enabling fence floor for Copter (logically) and Rover (potentially?).

> .... also, please squash your commits up into logical changes. There are patches in here which are definitely "fixups" of previous commits. I do realise this is WIP :-)

Apologies, I've tried to nail everything on my first PR, but things were bound to get missed. Certainly can do that.

peterbarker commented:
The fence min-alt stuff.... floor seems a little tacked-on ATM.

Should we consider making `enable(bool value)` instead `enable(uint8_t mask)`?

(I actually have a related PR which would make that `enable(FenceMask mask)` here: https://github.com/ArduPilot/ardupilot/pull/15073 - still needs a bit of work.

joshanne commented:
> The fence min-alt stuff.... floor seems a little tacked-on ATM.

As tacked-on as it is, keep in mind that it was the goal to change the minimum amount of `AC_Fence` in order to support Plane and its existing capability. The implementation is mostly using the existing framework to do so. Also, not saying that it's without need of rework, hence reviews! :-) (I've just noticed I've missed at least two other points involving fence floors - I'll add that as well).
- [ ] Add fence floor calculation to `check_destination_within_fence`
- [ ] Add fence floor calculation to `get_breach_distance`

> Should we consider making `enable(bool value)` instead `enable(uint8_t mask)`?
> (I actually have a related PR which would make that `enable(FenceMask mask)` here: #15073 - still needs a bit of work.

Certainly steering away from `uint8_t`'s would be good, especially for reporting breach types, fence types and action types. I'll need to look at the related PR in depth a little more to see what your enable function looks like (I couldn't see it on first glance)

joshanne commented:
Commits logically squashed for adding of `AC_Fence` to `ArduPlane`. To address the confusion around leaving `geofence` in, I'll remove the geofence code in a future commit.

amilcarlucas commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -33,18 +33,19 @@ const AP_Param::GroupInfo AC_Fence::var_info[] = {
     AP_GROUPINFO("ACTION",      2,  AC_Fence,   _action,        AC_FENCE_ACTION_RTL_AND_LAND),
 
-    // @Param{Copter, Sub}: ALT_MAX
+    // @Param: ALT_MAX
```
Rover also probably will not want this

joshanne replied:
Is there a use case where Rover may be climbing a mountain-side and exceed an altitude that needs to RTL?
I do note that it was not in the `AP_GROUPINFO_FRAME` definition, but, safest to check.

amilcarlucas reviewed (approved):
This makes one of my PR's ( #7111 ) to add a floor fence to copter mostly obsolete.
One think I have in mine though is that it uses the down-facing lidar, if present, for the floor fence calculations.

joshanne commented:
> This makes one of my PR's ( #7111 ) to add a floor fence to copter mostly obsolete.
One think I have in mine though is that it uses the down-facing lidar, if present, for the floor fence calculations.

Thanks @amilcarlucas - I'll take a look at your implementation. Perhaps this could be a two part process when the down-facing lidar is added as a follow-up?

amilcarlucas commented:
Yes that would be great

amilcarlucas commented:
Some of the CI builds are failing, can you look into that?

peterbarker commented:
The fence min-alt stuff.... floor seems a little tacked-on ATM.

Should we consider making `enable(bool value)` instead `enable(uint8_t mask)`?

(I actually have a related PR which would make that `enable(FenceMask mask)` here: https://github.com/ArduPilot/ardupilot/pull/15073 - still needs a bit of work.

joshanne commented:
> Some of the CI builds are failing, can you look into that?

Ahh, rebase but missed one last commit to finish removing geofence code. Using the `build_ci.sh` script now to check there's no other issue.

peterbarker commented on ArduPlane/commands_logic.cpp:
```diff
@@ -131,20 +131,17 @@ bool Plane::start_command(const AP_Mission::Mission_Command& cmd)
+            plane.fence.enable(true);
+            gcs().send_text(MAV_SEVERITY_INFO, "Fence enabled");
+        } else if (cmd.p1 == 2) { // disable fence floor only
+            plane.fence.floor_enable(false);
```
Could we get `enable_floor()` and `disable_floor()` instead?

joshanne replied:
We can.

peterbarker commented on ArduPlane/commands_logic.cpp:
```diff
@@ -384,21 +381,16 @@ void Plane::do_land(const AP_Mission::Mission_Command& cmd)
-            gcs().send_text(MAV_SEVERITY_NOTICE, "Fence floor disabled (auto disable)");
-        }
+#if AC_FENCE == ENABLED
+    if (fence.auto_enabled() == AC_Fence::AutoEnable::ALWAYS_ENABLED) {
```
This should `switch`

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+#if AC_FENCE == ENABLED
+
+// fence_check - ask fence library to check for breaches and initiate the response
+// called at 1Hz
```
No it's not :-P

These comments always get out of date.  Just remove this "called at" bit.

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+    // fence pre-arm check actually checks if any fence has been breached
+    // that's not ever going to be true if we don't call check on AP_Fence while disarmed
+    if (!arming.is_armed())
+    {
```
Brace on preceding line

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    if (!fence.enabled())
+    {
```
Brace on preceding line; similarly elsewhere.

joshanne replied:
On a dev call it was mentioned there was a script that performs these checks? My google-fu is weak today, apparently.

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+        if ((fence.get_action() == FENCE_ACTION_GUIDED || fence.get_action() == FENCE_ACTION_GUIDED_THR_PASS || fence.get_action() == FENCE_ACTION_RTL) &&
+            (control_mode == &mode_guided || control_mode == &mode_avoidADSB))
+        {
+            set_mode(*previous_mode, ModeReason::GCS_COMMAND);
```
This `ModeReason` isn't even close :-)  Feel free to add to the enumeration

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+    {
+        // Switch back to the chosen control mode if still in
+        // GUIDED to the return point
+        if ((fence.get_action() == FENCE_ACTION_GUIDED || fence.get_action() == FENCE_ACTION_GUIDED_THR_PASS || fence.get_action() == FENCE_ACTION_RTL) &&
```
I wonder if this should go before the `is_armed()` check...

Also.... this (preexisting) line sucks.  I wonder if we should just look at `plane.control_mode_reason` and switch back if that was `FenceBreach`

A switch on get_action might also help.

joshanne replied:
Actually points at a bigger issue - that uses `FENCE_ACTION` as what's defined in `mavlink`, and not the `AC_Fence` versions `AC_FENCE_ACTION`. I'll update that accordingly.

The way I read the old implementation is "if we are in a guided or avoid adsb mode because of a fence breach, and the fence is switched off, return to the previous mode."

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+        // if the user wants some kind of response and motors are armed
+        const uint8_t fence_act = fence.get_action();
+        switch (fence_act)
+        {
```
Brace on previous line

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+            } else {
+                //return to fence return point, not a rally point
+                guided_WP_loc = {};
+                if (fence.get_return_altitude() > 0) {
```
I wonder why we're not using this in the "true" clause in the if statement here.

joshanne replied:
Again, another case of carrying across existing implementation.

`fence.get_return_altitude()` uses the fence parameter for the fence return point altitude, which can be specified separately to the RTL altitude, which is used in the "true" clause. Either side of the clause could be changed, but that's a change to existing behaviour.

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+
+                AC_PolyFenceItem return_point;
+                //! TODO: Determine if this works?
+                if (fence.polyfence().get_item(fence.polyfence().num_stored_items() - 1, return_point) &&
```
Check my previous PR where I synthesises a return point if we didn't have one.... basically the geometric centre (read "average location") of all of the fencepoints.  Which, yes, can be outside the fence....

joshanne replied:
Awesome, I've nabbed the `polyfence().get_return_point(...)` to find the geometric centre. Thanks. 😄

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+
+        AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_FENCE, LogErrorCode(new_breaches));
+    }
+    else if (orig_breaches)
```
remove vertical whitespace before and after.

peterbarker commented on ArduPlane/fence.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+}
+
+bool Plane::fence_stickmixing(void)
```
Don't really like how throttle nudging is exempted here.  Would be a lot cleaner if it wasn't - but also a behavioural change.

peterbarker commented on ArduPlane/system.cpp:
```diff
@@ -159,6 +159,11 @@ void Plane::init_ardupilot()
+
+    // init fence
+#if AC_FENCE == ENABLED
+    
```
Well, that was unexciting....

joshanne replied:
wow! Actually needs to init inside the guards.

peterbarker commented on ArduPlane/takeoff.cpp:
```diff
@@ -271,13 +271,9 @@ int8_t Plane::takeoff_tail_hold(void)
-            gcs().send_text(MAV_SEVERITY_INFO, "Fence enabled (autoenabled)");
-        }
+#if AC_FENCE == ENABLED
+    if(fence.auto_enabled() >= AC_Fence::AutoEnable::ALWAYS_ENABLED) {
```
No~

```suggestion
    if(fence.auto_enabled() == AC_Fence::AutoEnable::ALWAYS_ENABLED) {
```

... and add in any other values you need....

joshanne replied:
Updated, and likened with landing implementation - that is `switch` on specific auto conditions.

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -103,8 +131,34 @@ void AC_Fence::enable(bool value)
-        clear_breach(AC_FENCE_TYPE_ALT_MAX | AC_FENCE_TYPE_CIRCLE | AC_FENCE_TYPE_POLYGON);
+        clear_breach(AC_FENCE_TYPE_ALT_MIN | AC_FENCE_TYPE_ALT_MAX | AC_FENCE_TYPE_CIRCLE | AC_FENCE_TYPE_POLYGON);
+    } else {
+        // turn the floor back on if it had been off
```
Not making any sense of this comment.  Seems unconditional here....

joshanne replied:
As you said: 
> These comments always get out of date.

Carried across from the `geofence_set_enabled` function. Can remove.

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -103,8 +131,34 @@ void AC_Fence::enable(bool value)
+void AC_Fence::floor_enable(bool value)
+{
+    if (_floor_enabled && !value) {
+        //! TODO: Log?
```
Maybe an event?

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -103,8 +131,34 @@ void AC_Fence::enable(bool value)
+
+bool AC_Fence::present() const
+{
+    // A fence is present if it is enabled and there are points present
```
We could have many inclusion / exclusion circles which Plane should now honour - and `present` ought to be true then!

joshanne replied:
So, a fences' presence is only tied to whether one is uploaded?

ie. We can have a fence (inclusion/exclusion/both) uploaded, but not enabled.

joshanne replied:
I've updated this to check inclusion/exclusion fences for circle and poly being present.

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -254,6 +313,54 @@ bool AC_Fence::check_fence_alt_max()
+    _curr_alt = -_curr_alt; // translate Down to Up
+
+    // check if we are under the altitude fence
+    if(_curr_alt < _alt_min) {
```
```suggestion
    if (_curr_alt < _alt_min) {
```

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -254,6 +313,54 @@ bool AC_Fence::check_fence_alt_max()
+    // clear alt breach if present
+    if ((_breached_fences & AC_FENCE_TYPE_ALT_MIN) != 0) {
+        clear_breach(AC_FENCE_TYPE_ALT_MIN);
+        _alt_min_backup = 0.0f;
```
Do we need to zero these?

joshanne replied:
Implementation for `check_fence_alt_min` is a reworked `check_fence_alt_max`, which contains this same assignment. I'd have to confirm the conditions in which they backup fence is used.

I suspect if we dont, the following series may occur:

- Flying within fences.
- Breach floor. Set backup floor
- Breach backup floor. Set new backup floor (repeat `n` times)
- Return inside fence
- Breach `FENCE_ALT_MIN`, `(!is_zero(_alt_min_backup) && _curr_alt < _alt_min_backup)` is treated as and old breach, because our current alt is less than `ALT_MIN` but greater than the `alt_min_backup`.

peterbarker commented on libraries/AC_Fence/AC_Fence.cpp:
```diff
@@ -460,6 +586,9 @@ bool AC_Fence::sys_status_enabled() const
     if (_action == AC_FENCE_ACTION_REPORT_ONLY) {
         return false;
     }
+    if (!_enabled) {
```
OK, you might have some fights on your hands here.  I'm not *that* fussed at this point, but changing behaviour like this could get interesting :-)

joshanne replied:
Figured that might be the case. Who's best to give feedback here?

What defines enabled?
What defines present?
What defines healthy?

joshanne replied:
I've reverted this back to not include the enabled check here. I'm not actually sure where this gets rendered after it's reported out. 

My assumptions were:
 - `present` is `true` if a fence (inclusion/exclusion circle/poly) is present.
 - `enabled` is true when the fence is enabled.
 - `healthy` is true if not breached.

If this flag is being used to let the user know if the fence is enabled, how do we accurately do that without looking at the enabled flag itself? I may have turned a fence off with a switch, but this flag doesn't know that.

peterbarker commented on libraries/AC_Fence/AC_Fence.h:
```diff
@@ -12,19 +12,23 @@
 #define AC_FENCE_ALT_MIN_DEFAULT                    -10.0f  // default maximum depth in meters
+#define AC_FENCE_ALT_MIN_BACKUP_DISTANCE            20.0f   // after fence is broken we recreate the fence 20m further down
 #define AC_FENCE_CIRCLE_RADIUS_DEFAULT              300.0f  // default circular fence radius is 300m
-#define AC_FENCE_ALT_MAX_BACKUP_DISTANCE            20.0f   // after fence is broken we recreate the fence 20m further up
```
Random reordering?

joshanne replied:
Thought it made more sense to group the max/min defaults together by type, but can swap them to be grouped as originally.

peterbarker commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -1217,6 +1225,13 @@ bool AP_Arming::disarm(const AP_Arming::Method method)
 
+    AC_Fence *fence = AP::fence();
+    if (fence != nullptr) {
+        if(fence->auto_enabled() == AC_Fence::AutoEnable::ONLY_WHEN_ARMED) {
```
I'm not sure this follows - we may have enabled the fence with a switch, for example....

However... existing behaviour...

joshanne replied:
Agree, existing behaviour. Intention was to push the disarming from the higher layers of plane to a common area for all vehicles. 

It's certainly debatable, I would say if you have the enabled fence via switch then you wouldn't want to turn them off automatically, unless your `AUTOENABLE` param was explicitly set to do so.

peterbarker reviewed (changes requested):
PR is incorrect; this will appear in Plane 4.1

`FENCE_CHANNEL` should be migrated to use the existing RC_Option.  This could be done ahead of time as a distinct PR - there's parameter conversion involved, so it's non-trivial


autotests:
 - ensure we can auto-enable the fence when vehicle armed
 - should disallow arming if we are outside the fence when we auto-enable
 - need tests around MAV_CMD_DO_FENCE_ENABLE
 - minimum altitude fence breach
 - fence_ret_rall works
 - need tests for what happens when fences are disabled while we're breached and undertaking an action

This PR is absolutely fantastic.  So many red lines!

joshanne commented on ArduPlane/Parameters.h:
```diff
@@ -104,6 +104,7 @@ class Parameters {
         k_param_gps,
         k_param_autotune_level,
         k_param_rally,
+        k_param_fence,
```
Bugger, had hoped it wasn't critical - I'll bump that to the end.

joshanne commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -391,9 +407,6 @@ void Plane::update_GPS_10Hz(void)
         }
 
-        // see if we've breached the geo-fence
-        geofence_check(false);
```
I had assumed (perhaps incorrectly) that `AC_Fence` using data supplied by `AP::ahrs().get_relative_position_NE_origin(...)` would be providing adjusted data. 

I would need to look further into what happens, though.

peterbarker commented:
On Thu, 10 Sep 2020, James O wrote:

> On a dev call it was mentioned there was a script that performs these checks? My google-fu is weak today, apparently.

Yes, we have an auto-formatter.  But it won't just check your delta, it 
will do the entire file, which is kind of bad as you'll end up with lots 
of unrelated meaningless changes making review a chore.

It's ./Tools/CodeStyle/ardupilot-astyle.sh

joshanne commented:
> It's ./Tools/CodeStyle/ardupilot-astyle.sh

Thanks for the pointer to the script. I'll take a look and see if it helps at all. Could be more hassle than it's worth, but good for use in the future.

joshanne commented:
> Oh, also, alt_min may be <-100 for Subs :-)

According to: http://www.ardusub.com/developers/full-parameter-list.html
If the sub can have a min fence < -100, that's not evident in existing documentation.

> **FENCE_ALT_MIN: Fence Minimum Altitude**
> Minimum altitude allowed before geofence triggers
> Range: -100 100
> Increment: 1
> Units: m

joshanne commented:
Todo list (on top of unresolved discussion points):

**Autotests**
- [ ] ensure we can auto-enable the fence when vehicle armed
- [ ] should disallow arming if we are outside the fence when we auto-enable
- [ ] need tests around MAV_CMD_DO_FENCE_ENABLE
- [x] minimum altitude fence breach
- [ ] fence_ret_rall works
- [ ] need tests for what happens when fences are disabled while we're breached and undertaking an action

hamishwillee commented:
@peterbarker What will the fence actions look like after this is done? 
- Are you likely to need the two modes that use the deprecated MAV_CMD_NAV_FENCE_RETURN_POINT? (this is problematic because it isn't clear how these should work if rally points are defined).
- One of these has manual throttle passthrough? Is that something that you plan to "keep"

I'm looking at defining MAV_CMD_NAV_FENCE_BREACH_ACTION to set the breach action for the next fence in a plan (so you can have fences for each item in the plan if you want). I'd like to use [FENCE_ACTION](https://mavlink.io/en/messages/common.html#FENCE_ACTION) but it is not very "common". 

Is this something we could change? If not, suggest we push this into ArduPilot dialect again and come up with a version that is more generic. Discussion is here: https://github.com/mavlink/mavlink/pull/1482

joshanne commented:
@hamishwillee As a rough indication, the intention is to maintain the same fence breach logic from `ArduPilot/Plane` in the fence library - this includes the throttle passthrough. As for the Fence Return Point, the initial cut intends to include the fence return point section, again to keep the logic the same.

That said, any architectural questions would be best directed to the plane devs.

amilcarlucas commented:
Should this be discussed at the devCall to speed things up ?

joshanne commented:
> Should this be discussed at the devCall to speed things up ?

I'm out of office for the AEST dev call (Tuesday morning), but can join for Wednesday's EU Dev Call. I too would be interested in agreeing what is required to speed this up.

rmackay9 commented on ArduPlane/Attitude.cpp:
```diff
@@ -58,7 +58,9 @@ bool Plane::stick_mixing_enabled(void)
         if (g.stick_mixing != STICK_MIXING_DISABLED &&
             g.stick_mixing != STICK_MIXING_VTOL_YAW &&
-            geofence_stickmixing() &&
+            #if AC_FENCE == ENABLED
```
Probably not a blocker but we often put the #if left justified..

joshanne replied:
Ack, also removed `#if` in `if` statement.

rmackay9 commented on ArduPlane/commands_logic.cpp:
```diff
@@ -384,21 +381,22 @@ void Plane::do_land(const AP_Mission::Mission_Command& cmd)
-            gcs().send_text(MAV_SEVERITY_NOTICE, "Disable fence floor failed (autodisable)");
-        } else {
+#if AC_FENCE == ENABLED
+    switch(fence.auto_enabled()) {
```
nitpick: normally we have a space between the "switch" and the brackett

rmackay9 commented on ArduPlane/radio.cpp:
```diff
@@ -211,7 +211,11 @@ void Plane::read_radio()
-    if (g.throttle_nudge && channel_throttle->get_control_in() > 50 && geofence_stickmixing()) {
+    if (g.throttle_nudge && channel_throttle->get_control_in() > 50 
+        #if AC_FENCE == ENABLED
+        && fence_stickmixing()
```
it's a bit hard to read having just part of the if within the #if so I'd probably instead have a local variable just above the if.

i.e. something like this:

    #if AC_FENCE == ENABLED
        const bool stickmixing = fence_stickmixing();
    #else
        const bool stickmixing = true;
    #end 
    if (g.throttle_nudge && channel_throttle->get_control_in() > 50 && stickmixing) {

amilcarlucas commented:
@joshanne this now needs a rebase, if you want this to be on 4.1 release .....

amilcarlucas commented:
@joshanne I went ahead and fixed the conflicts on #16050

joshanne commented:
@amilcarlucas Getting back to the PR this week hopefully. It did need a few things addressed in autotest, iirc.

tridge commented:
@joshanne hopefully you can use the branch from @amilcarlucas to save some time in the rebase

amilcarlucas commented:
Some CI tests are failing. Can you check it out @joshanne ?

amilcarlucas commented:
Now it also needs a rebase :(

amilcarlucas commented:
I did  the rebase here: #16050
@tridge can you merge  #15568 so that this one can be more easily tested?

joshanne commented:
As discussed here: https://github.com/ArduPilot/ardupilot/pull/16050#issuecomment-755690232

This changes the fence enable and autoenable to be frame specific, as well as certain other features of planes geofence (eg. min fence/floor, etc). I ran into roadblocks that appear to be more outside the scope of adding support for AC_Fence to plane, and as such they were left as frame specific params.

I believe the added autotests also cover these applied changes.

joshanne commented:
Strangely, the `test plane / build (base)` appears to have failed? I can't see a retry button for them - I assume it's just a small hiccup?

```
  Error: fatal: unable to access 'https://github.com/UAVCAN/pyuavcan/': Failed to connect to github.com port 443: Connection timed out
  Fetched in submodule path 'modules/uavcan/libuavcan/dsdl_compiler/pyuavcan', but it did not contain c58477a644d20ccf95a20c151f3a0402f271c3b8. Direct fetching of that commit failed.
  Submodule path 'modules/waf': checked out '67b3eac550520cd8a7cd1dce0f77820e91fb8ad0'
  Failed to recurse into submodule path 'modules/uavcan'
  Error: The process '/usr/bin/git' failed with exit code 1
```

joshanne commented:
Strangely, the `test plane / build (base)` appears to have failed? I can't see a retry button for them - I assume it's just a small hiccup?

```
  Error: fatal: unable to access 'https://github.com/UAVCAN/pyuavcan/': Failed to connect to github.com port 443: Connection timed out
  Fetched in submodule path 'modules/uavcan/libuavcan/dsdl_compiler/pyuavcan', but it did not contain c58477a644d20ccf95a20c151f3a0402f271c3b8. Direct fetching of that commit failed.
  Submodule path 'modules/waf': checked out '67b3eac550520cd8a7cd1dce0f77820e91fb8ad0'
  Failed to recurse into submodule path 'modules/uavcan'
  Error: The process '/usr/bin/git' failed with exit code 1
```

joshanne commented:
Looks like @amilcarlucas addressed some of the enable/disable fence floor issues - taking a look at it this morning.

joshanne commented:
@amilcarlucas `FRSkyMAVlite` is failing because it's expecting the MAV_CMD to enable fences is unsupported. The latest commit here fixes that.

I believe the test `DO_CHANGE_SPEED` is sporadic failure. Timeouts appear to be very tight, occasionally not reaching altitude within timeout, or not disarming within timeout. Locally, some runs, the test passes entirely.

amilcarlucas commented:
@joshanne Thanks for the fixes. This is starting to look very good.

amilcarlucas commented:
@joshanne can you give me permissions to write on your github repository, so that we can both work on this PR instead of having my forked branch?

peterbarker commented:
On Thu, 7 Jan 2021, Amilcar Lucas wrote:

> @joshanne can you give me permissions to write on your github repository, so that we can both work on this PR instead of having my forked branch?

You can PR into @joshanne's branch.  That will probably be a better way of 
approaching things....

joshanne commented:
> 
> 
> @joshanne can you give me permissions to write on your github repository, so that we can both work on this PR instead of having my forked branch?

I've just had a look, unfortunately I don't have the permissions in the organization to do this. I've pulled the branch to my personal repo too, I can give you permissions there, and manually sync back to the PR.

amilcarlucas commented:
OK, that is fine too. I'll close my PR #16050 now

amilcarlucas commented:
Can you give me permissions on your personal repo. I have some fixes.

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -976,17 +976,16 @@ def test_throttle_failsafe_fence(self):
+        }) # Turn fence on with aux function
+
+
+        self.mavproxy.expect("fence enabled")
```
better wait for MSG_FENCE_STATUS, it is send when fence is enable !

joshanne replied:
Ack - updated

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -976,17 +976,16 @@ def test_throttle_failsafe_fence(self):
+
+
+        self.mavproxy.expect("fence enabled")
+        self.delay_sim_time(1)  # This is needed else the SYS_STATUS may not have updated
```
if you need fo a new SYS_STATUS then just wait for it ! (self.mav.recv_match(type="SYS_STATUS", blocking=True, timeout=5)

joshanne replied:
We're not using a `SYS_STATUS` at this point - no need to delay sim - Removed.

khancyr commented on Tools/autotest/arducopter.py:
```diff
@@ -1298,6 +1299,48 @@ def fly_alt_max_fence_test(self):
 
+    # fly_alt_min_fence_test - fly down until you hit the fence
+    def fly_alt_min_fence_test(self):
+        self.takeoff(50, mode="LOITER", timeout=120)
```
just 30 will be enough, let's not waste to much time

khancyr commented on Tools/autotest/arducopter.py:
```diff
@@ -1298,6 +1299,48 @@ def fly_alt_max_fence_test(self):
+        self.set_parameter("FENCE_TYPE", 8)
+        self.set_parameter("FENCE_ALT_MIN", 20)
+
+        self.change_alt(50)
```
30

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -2062,6 +2053,53 @@ def change_speed():
+        self.progress("Fly above ceiling and check for breach")
+        self.change_altitude(self.homeloc.alt + cruise_alt + 80)
+        m = self.mav.recv_match(type='SYS_STATUS', blocking=True)
+        print("%s" % str(m))
```
self.progress

joshanne replied:
Only replying once, given there are several review comments to use `self.progress`.

I disagree this should be `self.progress` - This is not a progress statement, this is to print out the value returned by `self.mav.recv_match(type=....)`. This is also in line with many other test cases already present here.

Some examples of this occurring in multiple test files.:
* `arduplane.py`
  * `test_throttle_failsafe`
  * `test_throttle_failsafe_fence`
  * `test_adsb`
* `arducopter.py`
  * `test_manual_control`

peterbarker replied:
@joshanne sorry, @khancyr is correct here.

The other places you've found are.... inadequacies... in the code.  They tend to slip in every now and then.

joshanne replied:
... of course. Updated.

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -2062,6 +2053,53 @@ def change_speed():
+        self.change_altitude(self.homeloc.alt + cruise_alt)
+
+        m = self.mav.recv_match(type='SYS_STATUS', blocking=True)
+        print("%s" % str(m))
```
self.progress

joshanne replied:
See above

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -2062,6 +2053,53 @@ def change_speed():
+        self.change_altitude(self.homeloc.alt + cruise_alt - 80)
+
+        m = self.mav.recv_match(type='SYS_STATUS', blocking=True)
+        print("%s" % str(m))
```
self.progress

joshanne replied:
See above

khancyr reviewed (commented):
ideally we should test all option on autoenable fence to ensure that we can always land on failsafe or fence breach.

I am quite bothered by the auto enable thing and the floor fence. We need to ensure that it will be disable in case of failsafe unless specifically setted to not.

joshanne commented:
@amilcarlucas 
I've added tests for each Auto-enable condition (Disabled, AutoEnableAfterTakeoff, AutoEnableAfterTakeoffDisableFloorOnLand, OnArm). These auto-enable conditions are implemented in much the same way as the plane auto-enable conditions are implemented - that is when the takeoff can be verified in `guided` or `auto` modes. I'll need to make the same change across on sub, assuming it is acceptable this way.

amilcarlucas commented:
@joshanne parts of 5c9e8a5b90072a21dc5a85dc179f99305ac09f49 seam to be duplicated from 967d3c751ed46863e820e65b0312f1d2735238b6 Do you need that duplication?

joshanne commented:
If you're referring to `autoenable_floor_fence` it's more than likely that this function is not used after the changes I made. If you are autoenabling a fence, you want to enable the whole fence, and not just the floor (again, taking inspiration from plane here).

I'd propose that the entirety of that function be removed. It's also probable that plane uses a similar function, and so the auto-enabling of the fence on takeoff, and the auto-disabling of the fence/fence floor on landing could be made common in the AC_Fence library.

amilcarlucas commented:
yes, that is a good plan.

joshanne commented:
@amilcarlucas 
> yes, that is a good plan.

`auto_enable_fence_after_takeoff` and `auto_disable_fence_for_landing` are now common functions in the AC_Fence library.

There doesn't seem to be an appropriate place for `takeoff complete` or `landing start` in Sub or Rover, so I've updated the documentation to remove `AutoEnableOnTakeoff` and `AutoEnableDisableFloorOnLanding`. These vehicle frames can still use `3:AutoEnableOnlyWhenArmed` - auto enabling the fence when the vehicle is armed.

amilcarlucas commented:
Needs a rebase :(

joshanne commented:
@amilcarlucas 
> Needs a rebase :(

Rebased again.

amilcarlucas commented:
It would be very, very good if we can get this one in 4.1.

amilcarlucas commented:
@peterbarker this has a lot of tests now. Have you any ideas why the following test is failing:
Failure Summary:
  test.Plane:
    DO_CHANGE_SPEED (Test mavlink DO_CHANGE_SPEED command) (Initial speed not as expected (want=21.500000 got=19.604353) (see /tmp/buildlogs/ArduPlane-DO_CHANGE_SPEED.txt)
FAILED 1 tests: ['test.Plane']

joshanne commented:
> It would be very, very good if we can get this one in 4.1.

Here's hoping.

> this has a lot of tests now.

Re. Auto-tests, I can't see why that might be failing - running `DO_CHANGE_SPEED` locally seems to result in success.

peterbarker commented:
```
2021-01-19T23:44:41.1504643Z AT-0134.9: Attained Altitude=100.804000
2021-01-19T23:44:41.1505075Z RTL> fence breach
2021-01-19T23:44:41.1505578Z AT-0134.9: TAKEOFF COMPLETE
2021-01-19T23:44:41.1506255Z AT-0134.9: Entering guided and flying somewhere constant
```

There's a fence left over.

This commit might fix it: https://github.com/ArduPilot/ardupilot/pull/15141/commits/13226b516bb6fbafc83fd6a5075dfc08c29a5d7c (it clears fences and rally points between tests rather than just missions).

joshanne commented:
Ahh, thanks for pointing out that fence breach. Unfortunately, it's not a polygon breach, so clearing the polygon fences won't stop the breach. It actually looks like it's a tin-can fence breach (`BREACH_TYPE=2`). This is due to the fence being enabled via RC, and still enabled after executing `test_throttle_failsafe_fence`.

Tin-can is being breached at it's radius (300m), and aircraft RTLs. I'll add a fence disable to the end of the throttle failsafe fence test to ensure the fence has been returned to its previous state.

peterbarker commented:
On Thu, 21 Jan 2021, joshanne wrote:

> Ahh, thanks for pointing out that fence breach. Unfortunately, it's not a polygon breach, so clearing the polygon fences won't stop the breach. It
> actually looks like it's a tin-can fence breach (BREACH_TYPE=2). This is due to the fence being enabled via RC, and still enabled after executing
> test_throttle_failsafe_fence.

RC values should revert to 1500 - was it definitely enabled via RC?

> Tin-can is being breached at it's radius (300m), and aircraft RTLs. I'll add a fence disable to the end of the throttle failsafe fence test to ensure the
> fence has been returned to its previous state.

All parameters should also reset at the end ofa test - is the fence on and 
300m by default?!

The test framework makes a vague attempt at starting you in a sane 
environment and it seems to be failing in this case :-/

joshanne commented:
> RC values should revert to 1500 - was it definitely enabled via RC?

Ensured I only had two tests running, `test_throttle_failsafe_fence` followed by `fly_do_change_speed`.
`test_throttle_failsafe_fence` sets `RC7_OPTION` to Aux switch to enable fence. The point of this test appears to use the RC7 to enable the fence and ensure that it is still present when an RC failsafe occurs. The parameter `SIM_RC_FAIL` is then set to `2` , which resulted in in an RC7 channel that is always set to 1500 (even if manual attempts at setting it are made).

To overcome this, I have two options. 1) At the end of this test, run `self.do_fence_disable()`. 2) At the end of this test, remove the `SIM_RC_FAIL` condition, by setting it to zero, and setting RC7's value to 1000.

>All parameters should also reset at the end of a test - is the fence on and 300m by default?!

`FENCE_ENABLE` is not being set as a parameter in this case, so I suspect that is the reason it is not being reset at the end of the test. The fence is _not_ on by default - it is being left turned on after the throttle failsafe test has completed. The `AC_Fence` library as it currently stands has the tin-can fence being set to 300m by default (This value is not something added or changed).

peterbarker commented:
On Thu, 21 Jan 2021, joshanne wrote:

> To overcome this, I have two options. 1) At the end of this test, run self.do_fence_disable(). 2) At the end of this test, remove the SIM_RC_FAIL
> condition, by setting it to zero, and setting RC7's value to 1000.

Do whatever you need to do :-)

Remember that you need to worry about your test raising an exception - 
that's why we do so much of the "ex = e", cleanup and then raise ex again 
stuff.

I wouldn't mind reproducing this - can you drop a command line in, please?

joshanne commented:
> Remember that you need to worry about your test raising an exception - that's why we do so much of the "ex = e", cleanup and then raise ex again stuff.

Understood - the difference here is it's not a fail in this case... The test has successfully run, and is about to finish, but the vehicle is still in a failsafe condition.

> I wouldn't mind reproducing this - can you drop a command line in, please?

There's no command line option to run the two tests, I've simply modified `arduplane.py` in the `def tests(self):` to only add `ThrottleFailsafeFence` and `DO_CHANGE_SPEED`. It appears when you run `autotest.py fly.ArduPlane.ThrottleFailsafeFence fly.ArduPlane.DO_CHANGE_SPEED`, it runs them almost completely separately, so you won't see the error manifest.

I can add in a command line option and push it to the branch, but it's certainly a lot quicker to comment a handful of lines.

joshanne commented:
@peterbarker 
> I wouldn't mind reproducing this - can you drop a command line in, please?

looking through `autotest.py`, I don't see a way I can achieve a command line.

If you want to see the issue I was referring to, I've created a new branch with some autotest changes that allow you to run or skip specific tests of `test.Plane`.

https://github.com/joshanne/ardupilot/tree/autotest-run-specific-subtests-only

Use the command: `autotest.py build.Plane test.Plane --runsubtest ThrottleFailsafeFence,DO_CHANGE_SPEED`

To fix the issue, undo the comment made in 16779f92c93ec1b65cc394f274b18bee7674ba59. This disables the fence at the end of the test, and thus the two tests will pass.

amilcarlucas commented:
I guess that for subs the high altitude fence also needs to be "disabled" before "landing"

joshanne commented:
@amilcarlucas 
Looked at doing this some time ago [(here)](https://github.com/ArduPilot/ardupilot/pull/15288#issuecomment-759830991) - there doesn't appear to be a logical place for adding the auto-enable/disable to sub or rover.

Sub appears to only ever report a fence breach, and not change behaviour because of a fence breach. This is true at least in `ArduSub/fence.cpp`.

Current committed solution is to remove the ability for auto-enable/disable when 'takeoff' or 'landing', but allow other auto-enable conditions (eg. `When Armed`)

Davidsastresas commented:
I've just tested this PR on latest QGC master and it seems to work properly for Arduplane on SITL. Both inclusion and exclusion, cylindrical and polygon works as expected.

joshanne commented:
@Davidsastresas Thanks for your feedback, appreciate people looking at this.

Given someones looked at it, I'd be happy to discuss further in DevCall what next steps should be for it.

amilcarlucas commented:
I vote this for contribuition of the month. Is there anything else I should do to make it official candidate?

joshanne commented:
Latest changes here adds in a migration strategy for moving from geofence to AC_Fence - though it's still missing the mapping from geofence/FENCE_ACTION to AC_Fence/FENCE_ACTION. Any helpers??

Also adds:
* A new test for disabling the fence while performing the action, to ensure it returns to the correct mode (Static Test)
* A new test for RetRally flag, and whether the rally point or fence return point is used.
* Updates static fence tests to ensure the vehicle fails to arm if outside of a fence.

At this stage, I believe this is ready for full review.

joshanne commented:
Rebased and made `flake8` compliant.

amilcarlucas commented:
How is the wiki documentation coming?

joshanne commented:
> How is the wiki documentation coming?

No Wiki modifications have been made as yet. I can start putting something together.

joshanne commented:
@amilcarlucas Following up, there's been some work done for the wiki.

I've reached out to the _#documentation_ channel to seek clarity on a few points.

joshanne commented:
@amilcarlucas feel free to check out documentation PR (https://github.com/ArduPilot/ardupilot_wiki/pull/3349). It's about as complete as it can be without changes to GCS' to support AC_Fence in plane, and the additional params (eg. Alt_Min, Autoenable, etc) in other vehicle types.

amilcarlucas commented:
Looks good to me. Can you attend one of the Dev meetings this week so? Are you ready to remove the draft state on this?

joshanne commented:
I'll be present in both dev calls, happy to discuss at whichever has the shortest list, given how long they've been recently :-)

amilcarlucas commented:
This needs a rebase. Can you do it before the DevCall?

hamishwillee commented:
Howdy, is this implementation emitting FENCE_STATUS on message_interval request?

joshanne commented:
> This needs a rebase. Can you do it before the DevCall?

Rebased, I'll check back shortly and ensure builds passed.



> Howdy, is this implementation emitting FENCE_STATUS on message_interval request?

@hamishwillee `FENCE_STATUS` has been consolidated into GCS_MAVLink. I'm unsure how to validate what you're asking, but if it exists in copter/rover implementations, then it should also exist on this update for plane.

tridge commented:
testing:
 - need to disable tincan by default
 - default FENCE_TYPE should be polygon?
 - in upgrade handle alt max/min to set type
 - restore fence auto-enabled msg

tridge commented on ArduPlane/Parameters.cpp:
```diff
@@ -1461,6 +1412,50 @@ void Plane::load_parameters(void)
+                case 1: // FENCE_ACTION_GUIDED
+                    fence_action_new_val = AC_FENCE_ACTION_GUIDED;
+                    break;
+                case 3: // FENCE_ACTION_GUIDED_THR_PASS
```
missing RTL mode?

joshanne commented:
@tridge 

>     * need to disable tincan by default
>     * default FENCE_TYPE should be polygon?
>     * in upgrade handle alt max/min to set type
>     * restore fence auto-enabled msg
>     * missing RTL mode?

These issues have been addressed. I can't seem to find where your comment went, but it also addresses the issue of re-triggering a fence breach when jumping out of the `FENCE_ACTION` mode before fence breach is cleared. (ie. Cruise well beyond the fence wall, enable fence, re-enter cruise mode)

Also, made `FENCE_AUTOENABLE` a plane specific parameter, but left the ability to use fence floor in. Copter auto-enabling the floor when at RTL altitude should be a follow up change. However, the fence floor can still be used by enabling the fence with an aux switch `RCx_OPTION=11`, a mission item, or a mavlink cmd.

joshanne commented:
I've now also added an autotest to ensure mode change outside of the fence re-triggers the fence action (as per discovered during testing).

Also added an autotest to capture, and ensure the correctness of the overflow found in `get_return_point` when no return point is present. This occurs when longitudes are close to the max limits `[-180, 180]` and if in this state, could result in a fly away. eg. Longitudes close to CMAC, when added together exceed the `int32_t` size, overflow to negative, and half to be a longitude of `~-65`, placing the guided location somewhere near South America.

amilcarlucas commented on libraries/AC_Fence/AC_PolyFence_loader.cpp:
```diff
@@ -1158,8 +1158,8 @@ bool AC_PolyFence_loader::get_return_point(Vector2l &ret)
 
-    ret.x = ((min_loc.x+max_loc.x)/2);
-    ret.y = ((min_loc.y+max_loc.y)/2);
+    ret.x = (min_loc.x/2) + (max_loc.x/2);
```
Please add a comment. Someone reading these lines latter may ask why the strange math here. :)

joshanne commented:
@amilcarlucas comments added to `int32_t` overflow fix

peterbarker commented:
We should probably make sure that fence behaviour is unchanged when we lose our GPS (and other absolute positioning) sources.  This can be tested in SITL using `SIM_GPS_DISABLE`

amilcarlucas commented:
Many Thanks and Congrats to @joshanne . Great work!
